services/mirror/mirror_push.go

```diff
@@ -146,11 +147,33 @@ func runPushSync(ctx context.Context, m *repo_model.PushMirror) error {
 
 		log.Trace("Pushing %s mirror[%d] remote %s", path, m.ID, m.RemoteName)
 
+		var initArgs []string
+		if m.PublicKey != "" {
+			f, err := os.CreateTemp(os.TempDir(), m.RemoteName)
+			if err != nil {
+				log.Error("CreateTemp: %v", err)
+				return errors.New("unexpected error")
+			}
+			defer func() {
+				f.Close()
+				//if err := os.Remove(f.Name()); err != nil {
+				//	log.Error("os.Remove: %v", err)
+				//}
+			}()
+
+			if _, err := f.Write([]byte(m.PrivateKey)); err != nil {
+				log.Error("f.Write: %v", err)
+				return errors.New("unexpected error")
+			}
+
+			initArgs = append(initArgs, "-c", fmt.Sprintf("core.sshcommand=ssh -i %s -o IdentitiesOnly=yes", f.Name()))
```

Comment:
Maybe it's too early to review, but it seems that the `f.Name()` should be escaped.

For example, if `f.Name()` is `|| echo ...`, the `core.sshcommand` becomes `ssh -i || echo ... -o IdentitiesOnly=yes`

Just FYI 😊

Reply:
That's *luckily* not possible, go's `os/cmd` package already escapes them(magically, see https://github.com/go-gitea/gitea/pull/18835/commits/25bd4dbf905bab684cec4c1604d1a7ede4e7f7ae), Each item of the arguments slice would be considered as it's own argument. So it's already escaping on it's own. I will have to dig deeper to make sure, but escaping has only caused issues so far.

Reply:
`That's luckily not possible, go's os/cmd package already escapes them`: it is for the whole command `"core.sshcommand=ssh -i %s -o IdentitiesOnly=yes"`, but not for the embedded `core.sshcommand`.

IMO Git would run core.sshcommand directly.

Reply:
For what's worth it, it does seem to be possible to escape the file name. But the file name is entirely generated and doesn't contain user-input. So this would be preventive attack for a possible future use-case. https://github.com/go-gitea/gitea/blob/main/routers/web/repo/setting.go#L349

Reply:
My concern is that now it works, but in future someone else may reuse the code and pass user input into it, then attackers can execute any commands in Gitea (the most dangerous ...). In my mind, escaping everything correctly is the baseline for security.

Reply:
https://github.com/go-gitea/gitea/pull/18835/commits/ed6e4f803a19b5fcd5a01dab40e9402615dfff13#diff-12aceb8683cfc5ecf0718af8170844b1c9c48707f9704d9325a3ae5713028bcfR169

Reply:
I just realized that escaping arguments is more difficult than I thought ....

https://github.com/golang/go/blob/master/src/syscall/exec_windows.go#L27

Reply:
I haven't a windows machine ready to pull up. On Linux it correctly works to have the path quoted.